src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7164,12 +7164,63 @@ bool canFollowNullableType()
                     case ParseTypeMode.AfterIs:
                     case ParseTypeMode.DefinitePattern:
                     case ParseTypeMode.AsExpression:
-                        // These contexts might be a type that is at the end of an expression. In these contexts we only
-                        // permit the nullable qualifier if it is followed by a token that could not start an
-                        // expression, because for backward compatibility we want to consider a `?` token as part of the
-                        // `?:` operator if possible.
-                        //
-                        // Similarly, if we have `T?[]` we do want to treat that as an array of nullables (following
+                        // We are currently after `?` token after a type pattern
+                        // and need to decide how to parse next, e.g. `obj is string?| s ...` - `|` represents our position
+                        // There are 2 ways of treating identifier after `?` token after a type pattern:
+                        // 1. As a start of conditional expression, e.g. `var a = obj is string ? a : b`
+                        // 2. As a designation of a nullable-typed pattern, e.g. `if (obj is string? str)`
+                        // Since nullable types (no matter reference of value types) are not valid in patterns
+                        // by default we are biased towards the first option and consider case 2 only for error recovery purposes
+                        // (if we parse here as nullable type pattern an error will be reported during binding).
+                        // This condition checks for simple cases, where we better use option 2 and parse a nullable-typed pattern
+                        if (this.CurrentToken.Kind == SyntaxKind.IdentifierToken)
+                        {
+                            // Given that we are deciding whether we take `?:` operator or not while looking at state after `?`,
+                            // our focus is on `:` token as well. But `:` have special meaning in context of switch statements
+                            // and expressions (since `:` in switch expressions is recovered to `=>`). So consider the following case:
+                            // `case T? t:`. On its own `T ? t :` can be treated as a conditional expression with condition `T`,
+                            // 'whenTrue' `t` and 'whenFalse' not yet parsed. But it is way better to take `:` as case label end and
+                            // thus parse label condition as a declaration pattern `T? t`
+                            if (isTopLevelSwitchPattern)
```

Comment:
using hte word 'pattern' here was ery confusing. especially as it really seems like it's saying if we're parsign a swtich-statement or switch-expression arm. how about inSwitchStatementOrExpressionArmClause